Add support for Micrometer's Observation API for the SQS pipeline

sondemar opened the pull request:
## :loudspeaker: Type of change
<!--- Put an `x` in the boxes that apply -->
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring


## :scroll: Description
Added Micrometer's Observability support instruments SQS processing out of the box, while additionally propagating the inbound tracing information by looking it up in MessageHeaders.
Closes https://github.com/awspring/spring-cloud-aws/issues/646


## :bulb: Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->


## :green_heart: How did you test it?


## :pencil: Checklist
<!--- Put an `x` in the boxes that apply -->
- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] I updated reference documentation to reflect the change
- [x] All tests passing
- [x] No breaking changes


## :crystal_ball: Next steps

tomazfernandes commented:
Hi @sondemar, on a first glance this looks great!

I wonder if it'd also make sense to add Observability support to SqsTemplate? Or perhaps leave it to a separate issue / PR?

Let me know your thoughts, thanks!

sondemar commented:
> Hi @sondemar, on a first glance this looks great!
> 
> I wonder if it'd also make sense to add Observability support to SqsTemplate? Or perhaps leave it to a separate issue / PR?
> 
> Let me know your thoughts, thanks!

Hi @tomazfernandes , I have also thought about templates (SqsTemplate and SnsTemplate as well). It makes sense to include them because we might want to propagate the tracing context while sending messages to SQS/SNS or receiving messages from SQS (SqsTemplate). Initially, I considered including support for SQS/SNS templates in a separate ticket, but including SqsTemplate with this ticket is also fine.

tomazfernandes commented:
> Hi @tomazfernandes , I have also thought about templates (SqsTemplate and SnsTemplate as well). It makes sense to include them because we might want to propagate the tracing context while sending messages to SQS/SNS or receiving messages from SQS (SqsTemplate). Initially, I considered including support for SQS/SNS templates in a separate ticket, but including SqsTemplate with this ticket is also fine.

Makes sense, always a good idea to break into smaller PRs. In this case, TBH I'm not super familiar with Micrometer 2, so it'd be helpful to have a PR open for `SqsTemplate` instrumentation as well so I can run them locally and see the propagation happening, if that's ok with you.

tomazfernandes commented:
Another point here is - the most challenging part of instrumenting observation for this project is that it has many points where thread hopping may occur.

The main reason is it supports any combination of Blocking and Non-blocking components such as Interceptors, Message Listener and Error Handlers.

So I believe ideally we would have an Integration Test such as e.g.:
- An Interceptor that performs an async action using `AsyncSqsClient` (that should switch to an AWS SDK thread)
- A blocking interceptor that doesn't really have to do anything (that should switch to a thread from the standard / provided Task executor)
- An async message listener that performs an action using `AsyncSqsClient`  and throws an error (should switch to a AWS SDK Thread)
- A blocking error handler (should switch back to a `TaskExecutor` thread)

As a user I would expect to have Observability context in all these points, regardless of async / sync components.

Of course, we could start supporting only let's say the message listener, but then we'd need to make it clear in the docs what are the limitations.

Makes sense? Please let me know your thoughts. Thanks!

tomazfernandes commented:
Hello @sondemar, just to catch up on this

> but now I am waiting for feedback on the following questions: https://github.com/micrometer-metrics/context-propagation/issues/262 and https://github.com/micrometer-metrics/tracing/issues/765.

I see you're still looking into this, anything else we could do on this end?

Thanks!

sondemar commented:
> Hello @sondemar, just to catch up on this
> 
> > but now I am waiting for feedback on the following questions: [micrometer-metrics/context-propagation#262](https://github.com/micrometer-metrics/context-propagation/issues/262) and [micrometer-metrics/tracing#765](https://github.com/micrometer-metrics/tracing/issues/765).
> 
> I see you're still looking into this, anything else we could do on this end?
> 
> Thanks!

Hi @tomazfernandes !

I have found a solution for the issues I encountered. I am currently completing the integration tests proposed in [this comment](https://github.com/awspring/spring-cloud-aws/pull/1164#issuecomment-2198300841). I expect to have the full functionality ready by next week.

Thank you for your patience!

sondemar commented:
> #646

Hi @tomazfernandes ,

I have just pushed the following changes:

1.  Added observability for the SQS listener pipeline.
2.  Introduced observability for SqsTemplate.
3.  Created a test for your proposed scenario in [this comment](https://github.com/awspring/spring-cloud-aws/pull/1164#issuecomment-2198300841).

sondemar commented:
> I have just one question. This will send the trace when we send sns messages and we have SQS subscribing the SNS?

Hi @AnakinPt , According to [this comment](https://github.com/awspring/spring-cloud-aws/pull/1164#issuecomment-2317102908), the `SnsTemplate` is not included in this PR. As I mentioned previously in [this earlier comment](https://github.com/awspring/spring-cloud-aws/pull/1164#issuecomment-2170765175): 

> "Initially, I considered including support for SQS/SNS templates in a separate ticket, but including `SqsTemplate` with this ticket is also fine."

gtiwari333 commented:
Any idea on when/if this will get merged/released? Also, is there any workaround until we have this in place?

frjonsen commented:
We would also be very interested in seeing this merged. Is there anything blocking this PR?

tomazfernandes commented:
Hey @frjonsen, we're actively working on Observability and should have good news on this front in the next few weeks.

Stay tuned!

tomazfernandes commented:
Hey @sondemar, as I wrote in the other issue, I appreciate your PR very much but chose to go with a different approach.

Thanks!

Superseded by https://github.com/awspring/spring-cloud-aws/pull/1369
